runtime/src/main/cpp/Worker.cpp

```diff
@@ -373,34 +371,35 @@ void* workerRoutine(void* argument) {
   g_currentWorkerId = worker->id();
   Kotlin_initRuntimeIfNeeded();
 
-  while (true) {
-    Job job = worker->getJob();
-    if (job.function == nullptr) {
-       // Termination request, notify the future.
-      job.future->storeResultUnlocked(nullptr, true);
-      theState()->removeWorkerUnlocked(worker->id());
-      break;
-    }
+  {
     ObjHolder argumentHolder;
-    KRef argument = AdoptStablePointer(job.argument, argumentHolder.slot());
-    // Note that this is a bit hacky, as we must not auto-release resultRef,
-    // so we don't use ObjHolder.
-    // It is so, as ownership is transferred.
-    KRef resultRef = nullptr;
-    KNativePtr result = nullptr;
-    bool ok = true;
-    try {
-        job.function(argument, &resultRef);
+    ObjHolder resultHolder;
+    while (true) {
+      Job job = worker->getJob();
+      if (job.function == nullptr) {
+        // Termination request, notify the future.
+        job.future->storeResultUnlocked(nullptr, true);
+        theState()->removeWorkerUnlocked(worker->id());
+        break;
+      }
+
+      KRef argument = AdoptStablePointer(job.argument, argumentHolder.slot());
+      KNativePtr result = nullptr;
+      bool ok = true;
+      try {
+        job.function(argument, resultHolder.slot());
         argumentHolder.clear();
         // Transfer the result.
-        result = transfer(resultRef, job.transferMode);
-    } catch (ObjHolder& e) {
+        result = transfer(resultHolder.obj(), job.transferMode);
+      } catch (ObjHolder& e) {
         ok = false;
         if (worker->errorReporting())
             ReportUnhandledException(e.obj());
+      }
+      // Notify the future.
+      job.future->storeResultUnlocked(result, ok);
+      resultHolder.clear();
```

Comment:
As far as I see, by this moment another thread may have got this object already, so clearing the holder here may cause data race.

Btw, can `transfer` have `ObjHolder` parameter instead and clear it by itself? This looks making sense to me.